### PR TITLE
chore(explore): Get Explore data from endpoint instead of bootstrap_data

### PR DESCRIPTION
**Pull request by kgabryje**
### SUMMARY
This PR detaches Explore page from bootstrap data.
Getting `slice`, `form_data` and `datasource` is done by calling the new `/v1/explore` endpoint.
This is a prerequisite to Explore SPA project.

### TESTING INSTRUCTIONS
Everything should work like before (though a bit slower until we finish the rest of Explore work).
Make sure to test different initial urls (go to explore from charts list, dashboard, permalink...)

### ADDITIONAL INFORMATION

- [ ] Has associated issue:
- [ ] Required feature flags:
- [ ] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

### CODE REVIEW

File: superset-frontend/src/explore/actions/hydrateExplore.ts
Comment (michael-s-molina) [inline superset-frontend/src/explore/actions/hydrateExplore.ts:49]:
```suggestion
      // which will be manipulable by future reducers.
```

File: superset-frontend/src/explore/exploreUtils/getParsedExploreURLParams.test.ts
Comment (michael-s-molina) [inline superset-frontend/src/explore/exploreUtils/getParsedExploreURLParams.test.ts:30]: Maybe extract `http://localhost:9000/superset` to a const and reuse it in the tests?

File: superset-frontend/src/explore/actions/hydrateExplore.ts
Comment (michael-s-molina) [inline superset-frontend/src/explore/actions/hydrateExplore.ts:41]: The endpoint also returns a `message` attribute that contains information related to the request. One example is when we request `/explore` with an invalid `form_data_key`. In this case, `message` will contain `Form data not found in cache, reverting to chart metadata.` and this should be presented to the user in a toast.

**Review by michael-s-molina — COMMENTED** [review]
Thank you for the PR @kgabryje. I left some first-pass comments.

**Comment by jinghua-qa** [comment]
found a bug: blank page when create chart from the + button

https://user-images.githubusercontent.com/81597121/176377280-1572d8a4-ec8e-4ee7-a903-985e62c05b51.mov

File: superset-frontend/src/explore/actions/hydrateExplore.ts
Comment (zhaoyongjie) [inline superset-frontend/src/explore/actions/hydrateExplore.ts:43]: I think we should construct an appropriate/default formData here so that supports to open dataset.

```
    const initialFormData = slice?.form_data ? slice.form_data : form_data;
    const initialData = { form_data, slice, datasource };
    const initialControls = getControlsState(
      initialData,
      initialFormData,
    ) as ControlStateMapping;
    const { user, datasources, charts } = getState();

    const exploreState = {
....
....
      controls: initialControls,
....
    }
....
....
    const sliceFormData = slice
      ? getFormDataFromControls(initialControls)
      : null;
```

**Review by zhaoyongjie — COMMENTED** [review]
thanks for your hard work, leave some comments.

File: superset-frontend/src/explore/ExplorePage.tsx
Comment (kgabryje) [reply superset-frontend/src/explore/ExplorePage.tsx:41]: Good point!

**Comment by zhaoyongjie** [comment]
@kgabryje, explore page couldn't open from the dataset list, but worked when switching different datasets.

https://user-images.githubusercontent.com/2016594/176585320-c21418b6-904c-4653-aa13-0fafe4d0c0a8.mov

File: superset-frontend/src/explore/ExplorePage.tsx
Comment (zhaoyongjie) [inline superset-frontend/src/explore/ExplorePage.tsx:67]: Do you mind moving the HTTP request to a utility and returning a Promise object? This approach makes the data processing a little clearer.

File: superset-frontend/src/explore/actions/hydrateExplore.ts
Comment (zhaoyongjie) [inline superset-frontend/src/explore/actions/hydrateExplore.ts:52]: a nit for naming: the `initialData` should be `initialExploreState`

**Comment by zhaoyongjie** [comment]
@kgabryje @jinghua-qa I leave a few common use cases so that we can test this PR

1. Open an existing chart, success
2. Open an existing virtual/physical dataset, failure
3. Edit a Chart when clicking `edit chart` from Dashboard, success
4. Create a dataset from SQLLab, failure
5. Open an existing chart, and then switch to different datasets, success
6. Open an existing chart and the current user is admin; the user should permit editing the current dataset, failure
7. Manually construct an exception response, the explore page should show default interface, success

**Comment by jinghua-qa** [comment]
> @kgabryje @jinghua-qa I leave a few common use cases so that we can test this PR
> 
> #	Test case	Result
> 1	Open an existing chart	success
> 2	Open an existing virtual/physical dataset	failure
> 3	Edit a Chart when clicking `edit chart` from Dashboard	success
> 4	Create a dataset from SQLLab	failure
> 5	Open an existing chart, and then switch to different datasets	success
> 6	Open an existing chart and the current user is admin; the user should permit editing the current dataset	failure
> 7	Manually construct an exception response, the explore page should show default interface	success

Also see issue 2 and 4, issue 6 is current issue on master

**Comment by kgabryje** [comment]
Thank you so much @zhaoyongjie @jinghua-qa !  If the issues listed by Yongjie are already on master, let's address them in separate PRs.

**Comment by geido** [comment]
> > @kgabryje @jinghua-qa I leave a few common use cases so that we can test this PR
> > # Test case	Result
> > 1	Open an existing chart	success
> > 2	Open an existing virtual/physical dataset	failure
> > 3	Edit a Chart when clicking `edit chart` from Dashboard	success
> > 4	Create a dataset from SQLLab	failure
> > 5	Open an existing chart, and then switch to different datasets	success
> > 6	Open an existing chart and the current user is admin; the user should permit editing the current dataset	failure
> > 7	Manually construct an exception response, the explore page should show default interface	success
> 
> Also see issue 2 and 4, issue 6 is current issue on master

**Comment by geido** [comment]
> @kgabryje @jinghua-qa I leave a few common use cases so that we can test this PR
> 
> #	Test case	Result
> 1	Open an existing chart	success
> 2	Open an existing virtual/physical dataset	failure
> 3	Edit a Chart when clicking `edit chart` from Dashboard	success
> 4	Create a dataset from SQLLab	failure
> 5	Open an existing chart, and then switch to different datasets	success
> 6	Open an existing chart and the current user is admin; the user should permit editing the current dataset	failure
> 7	Manually construct an exception response, the explore page should show default interface	success

Wondering if we should add these cases to Cypress as these seems all scenarios that should have failed in CI.

P.S. Sorry didn't mean to close the PR :)

**Comment by kgabryje** [comment]
> Wondering if we should add these cases to Cypress as these seems all scenarios that should have failed in CI.

Agreed. I'll create a task for the next sprint

File: superset-frontend/src/explore/ExplorePage.tsx
Comment (kgabryje) [reply superset-frontend/src/explore/ExplorePage.tsx]: Done

**Review by zhaoyongjie — APPROVED** [review]
the main process test passed. continue to polish in the future.

**Comment by jinghua-qa** [comment]
> > @kgabryje @jinghua-qa I leave a few common use cases so that we can test this PR
> > # Test case	Result
> > 1	Open an existing chart	success
> > 2	Open an existing virtual/physical dataset	failure
> > 3	Edit a Chart when clicking `edit chart` from Dashboard	success
> > 4	Create a dataset from SQLLab	failure
> > 5	Open an existing chart, and then switch to different datasets	success
> > 6	Open an existing chart and the current user is admin; the user should permit editing the current dataset	failure
> > 7	Manually construct an exception response, the explore page should show default interface	success
> 
> Wondering if we should add these cases to Cypress as these seems all scenarios that should have failed in CI.
> 
> The code LGTM. I am going to pull this branch as I need it as a base to test another PR and I will report all eventual issues that I find, if any.
> 
> P.S. Sorry didn't mean to close the PR :)

Good idea!!

**Review by michael-s-molina — APPROVED** [review]
LGTM. Thank you for all the hard work @kgabryje! This is amazing and greatly simplifies the application. Thank you for addressing all the comments and bugs as well.

Let's just get QA's approval before merging it because of the high risk 😉

**Comment by codyml** [comment]
Tried all of the entrypoints to Explore I could think of and nothing broke!

**Review by geido — APPROVED** [review]
Code LGTM! Also manual testing didn't show any significant issue. With the OK of QA I think this is ready to go

**Review by jinghua-qa — APPROVED** [review]
QA team has check the related regression test and we did not see blocker issue. LGTM!
https://docs.google.com/spreadsheets/d/1aVEM4f4slnwAA-Bv9CMcQyQ9NSfT1vw5V0cXsNMmFUA/edit#gid=0

**Pull request merged**
